diff --git a/sites/all/modules/contrib/field_group/field_group.module b/sites/all/modules/contrib/field_group/field_group.module
index 412825f..4aff0d3 100644
--- a/sites/all/modules/contrib/field_group/field_group.module
+++ b/sites/all/modules/contrib/field_group/field_group.module
@@ -2258,7 +2258,13 @@ function field_group_field_group_to_hook_code($data, $module) {
   $code .= "  \${$export['identifier']}s = array();\n\n";
   foreach ($objects as $object) {
     $code .= ctools_export_object('field_group', $object, '  ');
-    $code .= "  \${$export['identifier']}s['" . check_plain($object->$export['key']) . "'] = \${$export['identifier']};\n\n";
+    if (defined('PHP_MAJOR_VERSION') && PHP_MAJOR_VERSION >= 7) {
+	// php 7
+	$code .= "  \${$export['identifier']}s['" . check_plain($object->{$export['key']}) . "'] = \${$export['identifier']};\n\n";
+    } else {
+	// php 5
+    	$code .= "  \${$export['identifier']}s['" . check_plain($object->$export['key']) . "'] = \${$export['identifier']};\n\n";
+    }
     if (!empty($object->data['label'])) {
       $translatables[] =  $object->data['label'];
     }
