diff --git a/field_group.module b/field_group.module
index 4fe2547..a0f18a9 100644
--- a/field_group.module
+++ b/field_group.module
@@ -2025,19 +2025,6 @@ function field_group_fields_nest(&$element, &$vars = NULL) {
       // Construct own weight, as some fields (for example preprocess fields) don't have weight set.
       $element[$group_name] = array();
       $group_references[$group_name] = &$element[$group_name];
-      // Get group parents
-      $parents = array();
-      $current_group = $group;
-      while (!empty($current_group)) {
-        array_unshift($parents, $current_group->group_name);
-        $current_group = isset($element['#fieldgroups'][$current_group->parent_name]) ?
-          $element['#fieldgroups'][$current_group->parent_name] : NULL;
-      }
-      $group_references[$group_name]['#array_parents'] = $parents;
-      $element['#fieldgroups'][$group_name]->array_parents = $parents;
-      // Remove self from parents and set #field_parents
-      array_pop($parents);
-      $group_references[$group_name]['#field_parents'] = $parents;
     }
   }
 
@@ -2088,19 +2075,6 @@ function field_group_fields_nest(&$element, &$vars = NULL) {
         // list intact (but if it is a field we don't mind).
         $group_references[$parent_name][$child_name] = &$element[$child_name];
         $group_references[$parent_name]['#weight'] = $element['#fieldgroups'][$parent_name]->weight;
-        // Prepend #array_parents & #field_parents of group child element & its element_children
-        // if those keys are set, and don't already include the group parents
-        $group_child = &$group_references[$parent_name][$child_name];
-        $group_parents = $group_references[$parent_name]['#array_parents'];
-        $process_elements = array_merge(array(&$group_child), _field_group_element_children_recursive_ref($group_child));
-        foreach ($process_elements as $key => $current_element) {
-          if (isset($current_element['#array_parents']) && !in_array($group_parents[0], $current_element['#array_parents'])) {
-            $process_elements[$key]['#array_parents'] = array_merge($group_parents, $current_element['#array_parents']);
-          }
-          if (isset($current_element['#field_parents']) && !in_array($group_parents[0], $current_element['#field_parents'])) {
-            $process_elements[$key]['#field_parents'] = array_merge($group_parents, $current_element['#field_parents']);
-          }
-        }
       }
 
       // The child has been copied to its parent: remove it from the root element.
@@ -2118,23 +2092,6 @@ function field_group_fields_nest(&$element, &$vars = NULL) {
 }
 
 /**
- * Recursive element_children, returns children by reference
- */
-function _field_group_element_children_recursive_ref(&$element) {
-  $results = array();
-  $children = element_children($element);
-  foreach ($children as $key) {
-    $child = &$element[$key];
-    if (is_array($child)) {
-      $results[] = &$child;
-      $results = array_merge($results, _field_group_element_children_recursive_ref($child));
-    }
-    unset($child);
-  }
-  return $results;
-}
-
-/**
  * Function to pre render the field group element.
  *
  * @see field_group_fields_nest()
