--- a/sites/all/modules/contrib/search_api_solr/includes/service.inc.orig
+++ b/sites/all/modules/contrib/search_api_solr/includes/service.inc
@@ -1725,7 +1725,8 @@ class SearchApiSolrService extends SearchApiAbstractService {
         $value = format_date($value, 'custom', self::SOLR_DATE_FORMAT, 'UTC');
         break;
       case 'text':
-        return '(' . call_user_func(array($this->connection_class, 'escape'), $value) . ')';
+       return call_user_func(array($this->connection_class, 'phrase'), $value); // massiveplatform
+        // return '(' . call_user_func(array($this->connection_class, 'escape'), $value) . ')';
     }
     return call_user_func(array($this->connection_class, 'phrase'), $value);
   }

